Bring back TypeGraphQL

Pl217 opened the pull request:
Most of the changes are cleanup and myself getting up to speed with the project and trying out the tools and libraries on Location data.

astroash committed:
🔥 Remove nexus dependency and generated artifacts

astroash committed:
➕ Add type-graphql

astroash committed:
🔥 Remove participants graphql types/resolver

astroash committed:
✨ Add plan types and resolver

Also, define `usageYear` table model and `planYear`
model, so that play years can be fetched.

Pl217 committed:
🔥 Remove `nexus` leftovers

Pl217 committed:
🔥 Remove `src/graphql-types.ts`

Pl217 committed:
🔥 Remove empty files under REST directory

Supporting REST endpoints has been dropped as an idea,
and files were just empty anyway

Pl217 committed:
➕ Add `typedi` dependency

This dependency is going to be used for as
inversion of control (dependency injection)
tool that lets us use decorator `@Service`
to inject business logic services into
GraphQL resolvers.

Pl217 committed:
🔧 Disable `strictPropertyInitialization`

This rule prevents writing TypeGraphQL object types
as initialization would be required by the compiler
and TypeGraphQL strictly prohibits writing constructors
in their object type modules.

Pl217 committed:
🩹 Import `reflect-metadata`

This package was installed, but it needs to be
explicitly imported at the top of entry file
in order for TS decorators to work.

Pl217 committed:
Use TypeGraphQL instead of nexus to create a schema

Pl217 committed:
Register `typedi` IOC container

This will enable injecting dependencies into
TypeGraphQL resolvers by using decorators.

Pl217 committed:
🐛 Fix knex `getOne` method

The method was generating invalid SQL

Pl217 committed:
🗃️ Define location DB models

Pl217 committed:
✨ Introduce location GraphQL types and resolvers

astroash commented on src/domain-services/base-types.ts:
```diff
@@ -0,0 +1,16 @@
+}
+
+@ObjectType()
+export class ParanoidType extends BaseType {
```
Is this concept borrowed from `sequalize`? (It's the only time I've come across soft delete being described in this way).

Personally I think that this is not descriptive naming, and would prefer something along the lines of `BaseTypeWithDeletedAt` or `BaseTypeWithSoftDelete`

Pl217 replied:
Yes, I've also seen this way of naming soft deletes in `sequelize` only. But used them anyway as I've seen https://github.com/UN-OCHA/hpc_service/pull/2241 enforce the term as well.

Renamed to `BaseTypeWithSoftDelete` which is indeed better name in 1f66d00ca0fdaf564406617dbe406f451e7279fa

astroash commented on src/domain-services/location/location-service.ts:
```diff
@@ -0,0 +1,23 @@
+import { LocationTable } from '../../data-providers/postgres/models/location';
+
+@Service()
+export class LocationService {
```
As we're moving back to `typegraphql`'s way of doing things, it is worth updating the file structure doc https://github.com/UN-OCHA/hpc-api/blob/develop/docs/FILE_STRUCTURE.md.

We previously were going to call our db abstractions libs, however `typedi` follows the spring structure of calling these services.

Pl217 replied:
Did some changes to mention services are used with help of dependency injection from `typedi` - see ba800a8aeb877bac6f4685d85793dadffd398d62.

astroash reviewed (changes requested):
Nice work tidying this up and adding in locations. 

Two small comments around naming conventions

Pl217 committed:
Use branded ID for location ID

Pl217 committed:
🔨 Promote `yarn` for development scripts

The repository is already having `yarn` over `npm`
in mind, so development scripts better reflect that.

Pl217 committed:
🏷️ Rename `ParanoidType` to `BaseTypeWithSoftDelete`

Pl217 committed:
📝 Update `docs/FILE_STRUCTURE.md`

Pl217 commented:
I am thinking that I should conclude my work on bringing back TypeGraphQL and complete the rest of points inside the [ticket](https://humanitarian.atlassian.net/browse/HPC-7904) with a new PR, to prevent this one from growing too much.

This mainly means adding authentication to requests, but that will also require adding some more types and resolvers.

astroash reviewed (approved):
These changes look great ✨
